src/coprocessor/dag/batch_executor/scan_executor.rs

```diff
@@ -0,0 +1,237 @@
+            // `self.scanner` should never be `None` when this function is being called.
+            unreachable!()
+        }
+    }
+
+    /// Get one row from the store.
+    #[inline]
+    fn point_get(&mut self, mut range: KeyRange) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+        let mut statistics = crate::storage::Statistics::default();
+        // TODO: Key and value doesn't have to be owned
+        let key = range.take_start();
+        let value = self.store.get(&Key::from_raw(&key), &mut statistics)?;
+        Ok(value.map(move |v| (key, v)))
+    }
+
+    /// Fills a column vector and returns whether or not all ranges are drained.
+    fn fill_column_vec(
+        &mut self,
+        expect_rows: usize,
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<bool> {
+        assert!(expect_rows > 0);
+
+        loop {
+            let range = self.ranges.next();
+            let some_row = match range {
+                super::ranges_iter::IterStatus::NewPointRange(r) => {
+                    self.scanned_rows_per_range.push(0);
+                    self.point_get(r)?
+                }
+                super::ranges_iter::IterStatus::NewNonPointRange(r) => {
+                    self.scanned_rows_per_range.push(0);
+                    self.reset_range(r)?;
+                    self.scan_next()?
+                }
+                super::ranges_iter::IterStatus::Continue => self.scan_next()?,
+                super::ranges_iter::IterStatus::Drained => {
+                    return Ok(true); // drained
+                }
+            };
+            if let Some((key, value)) = some_row {
+                // Retrieved one row from point range or non-point range.
+                self.scanned_rows_per_range
+                    .last_mut()
+                    .map_or((), |val| *val += 1);
+
+                self.imp.process_kv_pair(&key, &value, columns)?;
+
+                if columns.rows_len() >= expect_rows {
+                    return Ok(false); // not drained
+                }
+            } else {
+                // No more row in the range.
+                self.ranges.notify_drained();
+            }
+        }
+    }
+}
+
+/// Extracts `FieldType` from `ColumnInfo`.
+// TODO: Embed FieldType in ColumnInfo directly in Cop DAG v2 to remove this function.
+pub fn field_type_from_column_info(ci: &ColumnInfo) -> FieldType {
+    let mut field_type = FieldType::new();
+    field_type.set_tp(ci.get_tp());
+    field_type.set_flag(ci.get_flag() as u32); // FIXME: This `as u32` is really awful.
+    field_type.set_flen(ci.get_columnLen());
+    field_type.set_decimal(ci.get_decimal());
+    field_type.set_collate(ci.get_collation());
+    // Note: Charset is not provided in column info.
+    field_type
+}
+
+impl<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> BatchExecutor for ScanExecutor<S, I, P> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.imp.schema()
+    }
+
+    #[inline]
```

Comment:
The inline will happen for nested executors, for example, `SelectionExecutor<TableScanExecutor>`, eliminating a function call. In other cases I don't think the compiler is able to inline this function. Additionally the inline is just a hint and it does not force the compiler to inline the function.

Reply:
It confused me... So do you mean we can use `inline` everywhere since the compiler will decide to process it? Then when do we shouldn't use `inline`?

Reply:
@AndreMouche Not every function will be faster when inlined (due to cache locality) and not every function is worth to be inlined. Also there are disadvantages, e.g. losing call stacks, larger code base. So only these functions *worth* to be inlined should be marked as `#[inline]`. This function is an example that worth to be inlined because we **want** it to be inlined when there are nested executors (in the closure executor in future) to eliminate nested function calls. Consider a pipeline like TableScan - Selection - Limit - Aggregation. When they are nested in closure executor, we want **intermediate variables among nested `next_batch()` to be eliminated**. This can be achieved by inlining the function and compiler optimizations. This is the main reason that this function is marked as inline. In other cases, however, no inlining will happen because we call it behind dynamic dispatch (i.e. `Box<dyn BatchExecutor>`).

Reply:
ok